Support gdb load command

eulerdisk opened the pull request:
Closes #1057

Currently marked as draft because the flash operations extension in gdbstub PR is not merget yet:
(https://github.com/daniel5151/gdbstub/pull/172)
Maybe someone who's interested could test it meanwhile.

Some notes:
- I don't know if it make sense to reset after flash is done. Usually a `monitor reset` is necessary anyways.
- For the board I test flash sector info are wrong for OTP, option bytes etc.. That's not a problem for the load use case, but I wonder if there's some way to filter out those non-suited-for-exec-binary regions.

Yatekii commented:
This is very cool!

bugadani commented:
>  if there's some way to filter out those non-suited-for-exec-binary regions.

well behaved target descriptions properly set !executable for those regions. Each memory region has 4 flags: bootable, readable, writeable, executable, you should be able to check those.

There are a few questions not immediately obvious to me:
How do you actually load the image into memory? How do you handle the differences between image formats? How do you handle cases where parts of a flash block should be restored to the pre-erase value?

eulerdisk commented:
> well behaved target descriptions properly set !executable for those regions. Each memory region has 4 flags: bootable, readable, writeable, executable, you should be able to check those.

Thanks, I'll look into that in the next days.
 
> There are a few questions not immediately obvious to me: How do you actually load the image into memory? How do you handle the differences between image formats? How do you handle cases where parts of a flash block should be restored to the pre-erase value?

I'm not sure I understand your questions. GDB flash commands are kind of low-level and very similar to the `FlashLoader` API, I simply forward GDB commands to the `FlashLoader` object.
I've described how GDB `load` command works in the comments.

eulerdisk commented:
The new gdbstub version is up, so now the PR is now ready.
I also removed the last commit I added last week that filtered a certain type of memory region -- I don't think it's worth keeping.

Yatekii reviewed (approved):
Nice, thanks!
